src/analysis/asymptotics.lean

```diff
@@ -0,0 +1,438 @@
+by { rw ←is_littleo_norm_right, simp only [norm_neg], rw is_littleo_norm_right }
+
+theorem is_bigo_iff_pos {f : α → β} {g : α → γ} {l : filter α} :
+  is_bigo f g l ↔ ∃ c > 0, { x | ∥f x∥ ≤ c * ∥g x∥ } ∈ l.sets :=
+suffices is_bigo f g l → ∃ c > 0, { x | ∥f x∥ ≤ c * ∥g x∥ } ∈ l.sets,
+  from ⟨this, λ ⟨c, cpos, hc⟩, ⟨c, hc⟩⟩,
+assume ⟨c, hc⟩,
+or.elim (lt_or_ge 0 c)
+  (assume : c > 0, ⟨c, this, hc⟩)
+  (assume h'c : c ≤ 0,
+    have {x : α | ∥f x∥ ≤ 1 * ∥g x∥} ∈ l.sets,
+      begin
+        filter_upwards [hc], intros x,
+        show ∥f x∥ ≤ c * ∥g x∥ → ∥f x∥ ≤ 1 * ∥g x∥,
+        { intro hx, apply le_trans hx,
+          apply mul_le_mul_of_nonneg_right _ (norm_nonneg _),
+          show c ≤ 1, from le_trans h'c zero_le_one }
+      end,
+    ⟨1, zero_lt_one, this⟩)
+end
+
+section
+variables [normed_group β] [has_norm γ]
+
+@[simp]
+theorem is_bigo_norm_left {f : α → β} {g : α → γ} {l : filter α} :
+  is_bigo (λ x, ∥f x∥) g l ↔ is_bigo f g l :=
+by simp only [is_bigo, norm_norm]
+
+@[simp]
+theorem is_littleo_norm_left {f : α → β} {g : α → γ} {l : filter α} :
+  is_littleo (λ x, ∥f x∥) g l ↔ is_littleo f g l :=
+by simp only [is_littleo, norm_norm]
+
+@[simp]
+theorem is_bigo_neg_left {f : α → β} {g : α → γ} {l : filter α} :
+  is_bigo (λ x, -f x) g l ↔ is_bigo f g l :=
+by { rw ←is_bigo_norm_left, simp only [norm_neg], rw is_bigo_norm_left }
+
+@[simp]
+theorem is_littleo_neg_left {f : α → β} {g : α → γ} {l : filter α} :
+  is_littleo (λ x, -f x) g l ↔ is_littleo f g l :=
+by { rw ←is_littleo_norm_left, simp only [norm_neg], rw is_littleo_norm_left }
+
+theorem is_bigo.add {f₁ f₂ : α → β} {g : α → γ} {l : filter α}
+    (h₁ : is_bigo f₁ g l) (h₂ : is_bigo f₂ g l) :
+  is_bigo (λ x, f₁ x + f₂ x) g l :=
+let ⟨c₁, hc₁⟩ := h₁,
+    ⟨c₂, hc₂⟩ := h₂ in
+begin
+  use c₁ + c₂,
+  filter_upwards [hc₁, hc₂],
+  intros x hx₁ hx₂,
+  show ∥f₁ x + f₂ x∥ ≤ (c₁ + c₂) * ∥g x∥,
+  apply le_trans (norm_triangle _ _),
+  rw add_mul,
+  exact add_le_add hx₁ hx₂
+end
+
+theorem is_bigo.sub {f₁ f₂ : α → β} {g : α → γ} {l : filter α}
+    (h₁ : is_bigo f₁ g l) (h₂ : is_bigo f₂ g l) :
+  is_bigo (λ x, f₁ x - f₂ x) g l :=
+h₁.add (is_bigo_neg_left.mpr h₂)
+
+end
+
+section
+variables [normed_group β] [normed_group γ]
+
+theorem is_littleo_iff_pos {f : α → β} {g : α → γ} {l : filter α} :
+  is_littleo f g l ↔ ∀ c > 0, { x | c * ∥f x∥ ≤ ∥g x∥ } ∈ l.sets :=
+suffices (∀ c > 0, { x | c * ∥f x∥ ≤ ∥g x∥ } ∈ l.sets) → is_littleo f g l,
+  from ⟨λ h c cpos, h c, this⟩,
+assume h c,
+or.elim (lt_or_ge 0 c)
+  (assume : c > 0, h c this)
+  (assume h'c : c ≤ 0,
+    begin
+      filter_upwards [univ_mem_sets], intros x _,
+      show c * ∥f x∥ ≤ ∥g x∥,
+      { apply le_trans (mul_le_mul_of_nonpos_left (norm_nonneg _) h'c),
+        rw mul_zero, apply norm_nonneg }
+    end)
+
+theorem is_littleo.add {f₁ f₂ : α → β} {g : α → γ} {l : filter α}
+    (h₁ : is_littleo f₁ g l) (h₂ : is_littleo f₂ g l) :
+  is_littleo (f₁ + f₂) g l :=
```

Comment:
You used addition of functions here, but used lambda addition for `is_bigo.{add,sub}` and `is_littleo.sub`. I'm not sure which is better, but it should be consistent in any case. Probably the lambda addition will avoid some superfluous rewrites?

Reply:
You are right. Yes, I favored the lambdas, because they work better with rewrites. I'll fix this.